micro-rdk/src/common/data_manager.rs

```diff
@@ -0,0 +1,402 @@
+use std::time::Duration;
+
+use crate::common::data_collector::{DataCollectionError, DataCollector};
+use crate::common::data_store::DataStore;
+use crate::proto::app::data_sync::v1::SensorData;
+
+use super::data_collector::ResourceMethodKey;
+use super::data_store::{DataStoreError, WriteMode};
+use async_io::Timer;
+use bytes::BytesMut;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum DataManagerError {
+    #[error("no data collectors in manager")]
+    NoCollectors,
+    #[error(transparent)]
+    CollectionError(#[from] DataCollectionError),
+    #[error(transparent)]
+    StoreError(#[from] DataStoreError),
+}
+
+pub struct DataManager<StoreType> {
+    collectors: Vec<DataCollector>,
+    store: StoreType,
+    sync_interval_ms: u64,
+    part_id: String,
+}
+
+impl<StoreType> DataManager<StoreType>
+where
+    StoreType: DataStore,
+{
+    pub fn new(
+        collectors: Vec<DataCollector>,
+        store: StoreType,
+        sync_interval_ms: u64,
+        part_id: String,
+    ) -> Self {
+        Self {
+            collectors,
+            store,
+            sync_interval_ms,
+            part_id,
+        }
+    }
+
+    pub fn sync_interval_ms(&self) -> u64 {
+        self.sync_interval_ms
+    }
+
+    pub fn part_id(&self) -> String {
+        self.part_id.clone()
+    }
+
+    pub(crate) fn collection_intervals(&self) -> Vec<u64> {
+        let mut intervals: Vec<u64> = self.collectors.iter().map(|x| x.time_interval()).collect();
+        intervals.sort();
+        intervals.dedup();
+        intervals
+    }
+
+    pub async fn run(&mut self) -> Result<(), DataManagerError> {
```

Comment:
Just thinking out loud, but if this function ever does return a `DataManagerError`, what should be done with that error? Should this function get called again? Are we down forever?

Reply:
I believed that to be the responsibility of the task that initialized the manager and calls `run`? I could be convinced otherwise, maybe

Reply:
I'm fine with deferring an answer until later / higher up the stack.